gradle-conjure-api/src/main/java/com/palantir/gradle/conjure/api/ConjureProductDependenciesExtension.java

```diff
@@ -16,53 +16,37 @@
 
 package com.palantir.gradle.conjure.api;
 
-import groovy.lang.Closure;
-import groovy.lang.DelegatesTo;
-import java.util.HashSet;
 import java.util.Set;
-import javax.inject.Inject;
-import org.gradle.api.Project;
-import org.gradle.api.provider.ProviderFactory;
+import java.util.function.Function;
+import org.gradle.api.Action;
+import org.gradle.api.provider.Property;
 import org.gradle.api.provider.SetProperty;
 
-public class ConjureProductDependenciesExtension {
+public abstract class ConjureProductDependenciesExtension {
 
     public static final String EXTENSION_NAME = "serviceDependencies";
     public static final String ENDPOINT_VERSIONS_MANIFEST_KEY = "Endpoint-Version-Bounds";
 
-    private final Project project;
-    private final Set<ServiceDependency> productDependencies = new HashSet<>();
-    private final SetProperty<EndpointVersionBound> endpointVersions;
-    private final ProviderFactory providerFactory;
+    public abstract SetProperty<ServiceDependency> getProductDependencies();
 
-    @Inject
-    public ConjureProductDependenciesExtension(Project project) {
-        this.project = project;
-        this.endpointVersions =
-                project.getObjects().setProperty(EndpointVersionBound.class).empty();
-        this.providerFactory = project.getProviders();
-    }
+    public abstract SetProperty<EndpointVersionBound> getEndpointVersions();
 
-    public final Set<ServiceDependency> getProductDependencies() {
-        return productDependencies;
+    public ConjureProductDependenciesExtension() {
+        getProductDependenciesTransformer().convention(deps -> deps);
     }
 
-    public final void serviceDependency(@DelegatesTo(ServiceDependency.class) Closure<ServiceDependency> closure) {
-        ServiceDependency serviceDependency = new ServiceDependency();
-        closure.setDelegate(serviceDependency);
-        closure.call();
-        productDependencies.add(serviceDependency);
-    }
+    public abstract Property<Function<Set<ServiceDependency>, Set<ServiceDependency>>>
+            getProductDependenciesTransformer();
```

Comment:
I think we should allow multiple transformers on here. So it would be something like a `ListProperty<...>` that people `.add` to.

Reply:
Otherwise if we end up with two plugins trying to set this they will fight eachother.